### PR TITLE
[MINSTALL-115] Install At End feature (no extension)

### PR DESCRIPTION
**Pull request by cstamas**
This PR makes installAtEnd work as expected even
if maven-install-plugin is not used as extension.

How it works: it uses mojo Contexts to store "state markers":
* presence of marker means project was "processed"
* value of state marker tells what should be done

UTs adjusted to provide plugin context (was null before).
Plugin updated (as required by maven-plugin-testing-harness) and 
cleaned up.

---

https://issues.apache.org/jira/browse/MINSTALL-115

### CODE REVIEW

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 1]
```diff
@@ -91,53 +86,68 @@
...
-            synchronized ( INSTALLREQUESTS )
+            for ( MavenProject reactorProject : reactorProjects )
             {
-                while ( !INSTALLREQUESTS.isEmpty() )
+                Map<String, Object> pluginContext = session.getPluginContext( pluginDescriptor, reactorProject );
+                Boolean install = (Boolean) pluginContext.get( INSTALL_PROCESSED_MARKER );
+                if ( !install )
                 {
-                    installProject( session.getProjectBuildingRequest(), INSTALLREQUESTS.remove( 0 ) );
+                    getLog().info( "Project " + projectKey + " skipped install" );
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:118]: `projectKey` should be computed from `reactorProject` in this place

Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java]: fixed

**Comment by gnodet** [comment]
@cstamas This should not be affected by https://github.com/apache/maven/pull/625.  The reason is that when the session is cloned (this is already the case currently), the map containing the plugin contexts [1] is simply copied into the new session, so that they are actually shared.

[1] https://github.com/apache/maven/blob/31954b441b3cc9eabe3f8566669928cf91fe824a/maven-core/src/main/java/org/apache/maven/execution/MavenSession.java#L86-L87

File: pom.xml [hunk 2]
```diff
@@ -120,13 +132,31 @@
...
+    <dependency>
+      <groupId>org.eclipse.aether</groupId>
+      <artifactId>aether-util</artifactId>
+      <version>${aetherVersion}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.eclipse.aether</groupId>
+      <artifactId>aether-impl</artifactId>
+      <version>${aetherVersion}</version>
```
Comment (gnodet) [inline pom.xml:159]: It seems a bunch of changes are not really related to the `installAtEnd` problem and could be moved into a different commit.

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 3]
```diff
@@ -94,50 +89,70 @@ public void execute()
...
+            for ( MavenProject reactorProject : reactorProjects )
             {
-                while ( !INSTALLREQUESTS.isEmpty() )
+                Map<String, Object> pluginContext = session.getPluginContext( pluginDescriptor, reactorProject );
+                Boolean install = (Boolean) pluginContext.get( INSTALL_PROCESSED_MARKER );
+                if ( !install )
+                {
+                    getLog().info(
+                        "Project " + getProjectReferenceId( reactorProject ) + " skipped install"
+                    );
```
Comment (gnodet) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:118]: This looks redundant with the log statement printed earlier `"Skipping artifact installation"`.  This will print lots of line when the artifacts are actually installed for no benefit imho.

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 4]
```diff
@@ -94,50 +89,70 @@ public void execute()
...
-                    installProject( session.getProjectBuildingRequest(), INSTALLREQUESTS.remove( 0 ) );
+                    installProject( reactorProject );
                 }
             }
         }
         else if ( addedInstallRequest )
         {
-            getLog().info( "Installing " + project.getGroupId() + ":" + project.getArtifactId() + ":"
-                + project.getVersion() + " at end" );
+            getLog().info( "Installing " + getProjectReferenceId( project ) + " at end" );
```
Comment (gnodet) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:128]: Can this log statement be moved a bit earlier when `addedInstallRequest` is set to `true` ?

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 5]
```diff
@@ -94,50 +89,70 @@ public void execute()
...
             {
-                INSTALLREQUESTS.add( projectInstallerRequest );
+                getPluginContext().put( INSTALL_PROCESSED_MARKER, Boolean.TRUE );
                 addedInstallRequest = true;
             }
         }
 
-        boolean projectsReady = READYPROJECTSCOUNTER.incrementAndGet() == reactorProjects.size();
-        if ( projectsReady )
+        if ( allProjectsMarked() )
```
Comment (gnodet) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:108]: Does this work if the last project being built does not have a standard lifecycle ? Or maybe this does not exist ?

**Review by gnodet — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 2 above)
Comment (cstamas) [reply pom.xml:159]: Pulled out these into https://github.com/apache/maven-install-plugin/pull/17

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 5 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:108]: If m-install-p is not present, will not, but all "standard" life-cycle has them. But true, if no m-install-p not present, will not work. Was thinking about `org.apache.maven.execution.MavenSession#getResult` (and use `org.apache.maven.execution.MavenExecutionResult#getBuildSummary` to check "is it done"), wdyt?

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 3 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:118]: The message you refers to is when maven is invoked with `-Dmaven.install.skip`, when whole m-install-p is skipped.

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 5 above)
Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:107]: Neglect that above. Instead: this does NOT modify the existing feature (when it forced you to use plugin as extension), and will fix this in some newer PR

**Comment by cstamas** [comment]
@gnodet @michael-o ping, Guillaume, you have change requests here (unsure what). But let's agree all this PR does is:
* not changing existing behaviour (re last module)
* but removes the requirement for extension

So IMHO is def an improvement

**Comment by michael-o** [comment]
Will check on train tomorrow

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 6]
```diff
@@ -91,53 +86,68 @@
...
         {
-            synchronized ( INSTALLREQUESTS )
+            for ( MavenProject reactorProject : reactorProjects )
             {
-                while ( !INSTALLREQUESTS.isEmpty() )
+                Map<String, Object> pluginContext = session.getPluginContext( pluginDescriptor, reactorProject );
+                Boolean install = (Boolean) pluginContext.get( INSTALL_PROCESSED_MARKER );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:112]: Can this be `null` at any time?

Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:112]: no, as it is here, the `allProjectsMarked()` returned true, hence, all the modules has been marked (have value set)

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 7]
```diff
@@ -83,24 +83,30 @@
...
+    }
+
     public void execute()
         throws MojoExecutionException, MojoFailureException
     {
         if ( skip )
         {
-            getPluginContext().put( INSTALL_PROCESSED_MARKER, Boolean.FALSE );
             getLog().info( "Skipping artifact installation" );
+            getPluginContext().put( INSTALL_PROCESSED_MARKER, State.SKIPPED.name() );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:97]: If we have objects, why not retain the enum, why use name?

Comment (cstamas) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:97]: The span across several classloaders, there is no single enum.

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java [hunk 8]
```diff
@@ -88,56 +77,88 @@
...
+    private enum State
+    {
+        SKIPPED, INSTALLED, TO_BE_INSTALLED
+    }
+
+    private static final String INSTALL_PROCESSED_MARKER = InstallMojo.class.getName() + ".processed";
+
+    private void putState( State state )
+    {
+        getPluginContext().put( INSTALL_PROCESSED_MARKER, state.name() );
```
Comment (slawekjaranowski) [inline src/main/java/org/apache/maven/plugins/install/InstallMojo.java:89]: here `getPluginContext()` and in other lines, eg in `getState` method - `session.getPluginContext(...)` - why different method is used for state?

Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:89]: good question
because Mojo API has a direct `getPluginContext()` method to get the plugin context for the current project
but getting the plugin context for another project in the reactor is a different API that is provided in MavenSession: `MavenSession.getPluginContext( PluginDescriptor, MavenProject )`

File: src/main/java/org/apache/maven/plugins/install/InstallMojo.java (hunk 5 above)
Comment (hboutemy) [reply src/main/java/org/apache/maven/plugins/install/InstallMojo.java:126]: > if the last project being built does not have a standard lifecycle ?

is this a pure theoretical question (legit when thinking, but...)? Or do you know any concrete case?

IMHO, the best we can do is to add a note in plugins' parameter documentation

**Review by hboutemy — APPROVED** [review]

**Review by slawekjaranowski — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #276
